File: android/app/src/main/java/com/yesilai/app/viewmodel/ChatViewModel.kt
package com.yesilai.app.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import com.yesilai.app.data.model.ChatMessage
import com.yesilai.app.data.repository.ChatRepository
import com.yesilai.app.dataStore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import java.util.UUID

data class ChatUiState(
    val messages: List<ChatMessage> = emptyList(),
    val isLoading: Boolean = false,
    val isLoadingHistory: Boolean = true,
    val inputText: String = ""
)

class ChatViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = ChatRepository()
    private val sessionIdKey = stringPreferencesKey("session_id")
    
    private val _uiState = MutableStateFlow(ChatUiState())
    val uiState: StateFlow<ChatUiState> = _uiState.asStateFlow()
    
    private var sessionId: String = ""
    
    init {
        viewModelScope.launch {
            sessionId = getOrCreateSessionId()
            loadChatHistory()
        }
    }
    
    private suspend fun loadChatHistory() {
        _uiState.value = _uiState.value.copy(isLoadingHistory = true)
        
        val history = repository.loadChatHistory()
        
        if (history.isEmpty()) {
            // Add welcome message
            val welcomeMessage = ChatMessage(
                id = 1,
                text = "Merhaba! Bağımlılıkla mücadele yolculuğunda sana destek olmak için buradayım. Nelerden bahsetmek istersin?",
                sender = ChatMessage.MessageSender.BOT
            )
            _uiState.value = _uiState.value.copy(
                messages = listOf(welcomeMessage),
                isLoadingHistory = false
            )
            repository.saveMessage(welcomeMessage)
        } else {
            _uiState.value = _uiState.value.copy(
                messages = history,
                isLoadingHistory = false
            )
        }
    }
    
    private suspend fun getOrCreateSessionId(): String {
        val context = getApplication<Application>()
        val storedId = context.dataStore.data.map { preferences ->
            preferences[sessionIdKey]
        }.first()
        
        return storedId ?: run {
            val newId = "session_${System.currentTimeMillis()}_${UUID.randomUUID().toString().take(9)}"
            context.dataStore.edit { preferences ->
                preferences[sessionIdKey] = newId
            }
            newId
        }
    }
    
    fun updateInputText(text: String) {
        _uiState.value = _uiState.value.copy(inputText = text)
    }
    
    fun sendMessage() {
        val currentMessage = _uiState.value.inputText.trim()
        if (currentMessage.isEmpty() || _uiState.value.isLoading) return
        
        viewModelScope.launch {
            // Add user message
            val userMessage = ChatMessage(
                id = System.currentTimeMillis(),
                text = currentMessage,
                sender = ChatMessage.MessageSender.USER
            )
            
            _uiState.value = _uiState.value.copy(
                messages = _uiState.value.messages + userMessage,
                inputText = "",
                isLoading = true
            )
            
            // Save user message to Firestore
            repository.saveMessage(userMessage)
            
            // Send to API
            val result = repository.sendMessage(currentMessage, sessionId)
            
            val botResponseText = result.getOrElse { error ->
                when {
                    error.message?.contains("timeout", ignoreCase = true) == true ->
                        "⚠️ İstek zaman aşımına uğradı. Sunucu yanıt vermekte yavaş olabilir. Lütfen tekrar deneyin."
                    error.message?.contains("Unable to resolve host", ignoreCase = true) == true ->
                        "⚠️ Sunucuya ulaşılamıyor. İnternet bağlantınızı kontrol edin."
                    else ->
                        "⚠️ Bir sorun oluştu: ${error.message}"
                }
            }
            
            val botMessage = ChatMessage(
                id = System.currentTimeMillis() + 1,
                text = botResponseText,
                sender = ChatMessage.MessageSender.BOT
            )
            
            _uiState.value = _uiState.value.copy(
                messages = _uiState.value.messages + botMessage,
                isLoading = false
            )
            
            // Save bot message to Firestore
            repository.saveMessage(botMessage)
        }
    }
    
    fun testWebhook() {
        viewModelScope.launch {
            val testMessage = "Test mesajı - ${java.text.SimpleDateFormat("HH:mm:ss", java.util.Locale.getDefault()).format(java.util.Date())}"
            
            _uiState.value = _uiState.value.copy(isLoading = true)
            
            val result = repository.sendMessage(testMessage, sessionId)
            
            val responseText = result.getOrElse { "Test hatası: ${it.message}" }
            
            val testResponseMessage = ChatMessage(
                id = System.currentTimeMillis(),
                text = "🧪 Webhook Test: $responseText",
                sender = ChatMessage.MessageSender.BOT
            )
            
            _uiState.value = _uiState.value.copy(
                messages = _uiState.value.messages + testResponseMessage,
                isLoading = false
            )
            
            // Save test message to Firestore
            repository.saveMessage(testResponseMessage)
        }
    }
    
    fun clearChatHistory() {
        viewModelScope.launch {
            repository.clearChatHistory()
            
            val welcomeMessage = ChatMessage(
                id = 1,
                text = "Merhaba! Bağımlılıkla mücadele yolculuğunda sana destek olmak için buradayım. Nelerden bahsetmek istersin?",
                sender = ChatMessage.MessageSender.BOT
            )
            
            _uiState.value = _uiState.value.copy(messages = listOf(welcomeMessage))
            repository.saveMessage(welcomeMessage)
        }
    }
}

File: ios/YesilAI/ViewModels/ChatViewModel.swift
import Foundation
import SwiftUI
import FirebaseFirestore
import FirebaseAuth

@MainActor
class ChatViewModel: ObservableObject {
    @Published var messages: [ChatMessage] = []
    @Published var inputText: String = ""
    @Published var isLoading: Bool = false
    @Published var isLoadingHistory: Bool = false
    
    private let apiService = ChatAPIService.shared
    private let db = Firestore.firestore()
    
    private var userId: String? {
        Auth.auth().currentUser?.uid
    }
    
    init() {
        loadChatHistory()
    }
    
    func loadChatHistory() {
        guard let userId = userId else {
            addWelcomeMessage()
            return
        }
        
        isLoadingHistory = true
        
        db.collection("users").document(userId).collection("messages")
            .order(by: "timestamp", descending: false)
            .limit(to: 100)
            .getDocuments { [weak self] snapshot, error in
                guard let self = self else { return }
                
                Task { @MainActor in
                    self.isLoadingHistory = false
                    
                    if let error = error {
                        print("Error loading chat history: \(error.localizedDescription)")
                        self.addWelcomeMessage()
                        return
                    }
                    
                    guard let documents = snapshot?.documents, !documents.isEmpty else {
                        self.addWelcomeMessage()
                        return
                    }
                    
                    self.messages = documents.compactMap { doc -> ChatMessage? in
                        let data = doc.data()
                        guard let text = data["text"] as? String,
                              let senderRaw = data["sender"] as? String,
                              let sender = ChatMessage.MessageSender(rawValue: senderRaw),
                              let timestamp = data["timestamp"] as? Timestamp else {
                            return nil
                        }
                        
                        return ChatMessage(
                            id: Int64(timestamp.dateValue().timeIntervalSince1970 * 1000),
                            text: text,
                            sender: sender
                        )
                    }
                    
                    if self.messages.isEmpty {
                        self.addWelcomeMessage()
                    }
                }
            }
    }
    
    private func addWelcomeMessage() {
        let welcomeMessage = ChatMessage(
            id: 1,
            text: "Merhaba! Bağımlılıkla mücadele yolculuğunda sana destek olmak için buradayım. Nelerden bahsetmek istersin?",
            sender: .bot
        )
        messages = [welcomeMessage]
        saveMessage(welcomeMessage)
    }
    
    func sendMessage() {
        let messageText = inputText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !messageText.isEmpty, !isLoading else { return }
        
        // Add user message
        let userMessage = ChatMessage(
            id: Int64(Date().timeIntervalSince1970 * 1000),
            text: messageText,
            sender: .user
        )
        messages.append(userMessage)
        saveMessage(userMessage)
        
        inputText = ""
        isLoading = true
        
        Task {
            do {
                let response = try await apiService.sendMessage(messageText)
                let botMessage = ChatMessage(
                    id: Int64(Date().timeIntervalSince1970 * 1000) + 1,
                    text: response,
                    sender: .bot
                )
                messages.append(botMessage)
                saveMessage(botMessage)
            } catch {
                let errorMessage = ChatMessage(
                    id: Int64(Date().timeIntervalSince1970 * 1000) + 1,
                    text: error.localizedDescription,
                    sender: .bot
                )
                messages.append(errorMessage)
                saveMessage(errorMessage)
            }
            isLoading = false
        }
    }
    
    private func saveMessage(_ message: ChatMessage) {
        guard let userId = userId else { return }
        
        let messageData: [String: Any] = [
            "text": message.text,
            "sender": message.sender.rawValue,
            "timestamp": FieldValue.serverTimestamp()
        ]
        
        db.collection("users").document(userId).collection("messages")
            .addDocument(data: messageData) { error in
                if let error = error {
                    print("Error saving message: \(error.localizedDescription)")
                }
            }
    }
    
    func clearChatHistory() {
        guard let userId = userId else { return }
        
        // Delete all messages from Firestore
        db.collection("users").document(userId).collection("messages")
            .getDocuments { [weak self] snapshot, error in
                guard let documents = snapshot?.documents else { return }
                
                let batch = self?.db.batch()
                documents.forEach { batch?.deleteDocument($0.reference) }
                
                batch?.commit { _ in
                    Task { @MainActor in
                        self?.messages = []
                        self?.addWelcomeMessage()
                    }
                }
            }
    }
    
    func testWebhook() {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "HH:mm:ss"
        let testMessage = "Test mesajı - \(dateFormatter.string(from: Date()))"
        
        isLoading = true
        
        Task {
            do {
                let response = try await apiService.sendMessage(testMessage)
                let testResponseMessage = ChatMessage(
                    id: Int64(Date().timeIntervalSince1970 * 1000),
                    text: "🧪 Webhook Test: \(response)",
                    sender: .bot
                )
                messages.append(testResponseMessage)
                saveMessage(testResponseMessage)
            } catch {
                let errorMessage = ChatMessage(
                    id: Int64(Date().timeIntervalSince1970 * 1000),
                    text: "🧪 Test hatası: \(error.localizedDescription)",
                    sender: .bot
                )
                messages.append(errorMessage)
                saveMessage(errorMessage)
            }
            isLoading = false
        }
    }
}
